keps/sig-node/4576-conditions-for-pod-terminated/README.md

```diff
@@ -0,0 +1,918 @@
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests for meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes a
+`PodTerminated` condition in pod status to indicate a pod has
+reached a state where it's containers have terminated.
+
+## Motivation
+
+Today, there are two phases, `Succeeded` and `Failed`, which are used to indicate that
+all containers in the pod have been terminated.  
+The kubelet may update the phase of
+the pod to `Succeeded` or `Failed` in the [SyncPod](https://github.com/kubernetes/kubernetes/blob/dc3f5ec6ccb9855dfa99f4c1078625df5fdfab6a/pkg/kubelet/kubelet.go#L1723)
+and  waits for the corresponding resource dependencies to be cleanup in the [SyncTerminatedPod](https://github.com/kubernetes/kubernetes/blob/dc3f5ec6ccb9855dfa99f4c1078625df5fdfab6a/pkg/kubelet/kubelet.go#L2146)
+when all containers of the pod are terminated. In this case it is not safe to remove
+[pvc protection](https://github.com/kubernetes/kubernetes/issues/120756)
+because the phase of the pod already be in the `Succeeded` or `Failed` state 
+but the volumes are not unmounted(the umount process is asynchronous with the pod's phase change).
```

Comment:
I'm wondering if this issue is really about async unmounting of volumes, it it was I assume the pvc could be deleted soon after the pod is transitioned to `Failed`. However, it stays `Terminating` for a long time, so I don't see a clear connection between the issue and the condition.

Reply:
@mimowo PTAL at https://github.com/kubernetes/kubernetes/pull/123320.

Reply:
Because we don't have the right signals to determine when a volume has finished unmounting from the node, that means we cannot remove the finalizer protection.

Reply:
I might ask lame questions, as I'm not a sig-node expert - I just got interested in the KEP as it is close to the topics and code I was touching in the past. So, the question was just to build my understanding rather than pointing out flaws. 

Let me ask a follow up question in the similar spirit - what will happen if we remove the "finializer protection" while the pod is in terminal phase (say Failed), but the volumes are not unmounted yet? Would it be problematic?

Reply:
@mimowo Interesting, I'll do some testing.